# Enforces correct Tailwind CSS arbitrary value syntax (enforces-arbitrary-value-syntax)

This rule enforces the correct Tailwind CSS arbitrary value syntax, which is using square brackets `[]` to wrap arbitrary values.

## Rule Details

In Tailwind CSS, when you need to use a custom value that is not in the preset, you should use the arbitrary value syntax `prefix-[value]` instead of directly appending the value after the prefix.

### Incorrect Examples

```html
<div class="p-10px mx-15px leading-20px text-60px w-800px">
  These class names use the incorrect syntax
</div>
```

```js
classnames("p-10px", "mx-15px", "leading-20px", "text-60px", "w-800px");
```

### Correct Examples

```html
<div class="p-[10px] mx-[15px] leading-[20px] text-[60px] w-[800px]">
  These class names use the correct syntax
</div>
```

```js
classnames(
  "p-[10px]",
  "mx-[15px]",
  "leading-[20px]",
  "text-[60px]",
  "w-[800px]"
);
```

## Options

```js
...
"tailwindcss/enforces-arbitrary-value-syntax": [<enabled>, {
  "callees": Array<string>,
  "config": <string>|<object>,
  "skipClassAttribute": <boolean>,
  "tags": Array<string>,
}]
...
```

### `callees` (default: `["classnames", "clsx", "ctl", "cva", "tv"]`)

If you use tool libraries like [@netlify/classnames-template-literals](https://github.com/netlify/classnames-template-literals), you can add their names to the list to ensure that the rule can parse them correctly.

### `ignoredKeys` (default: `["compoundVariants", "defaultVariants"]`)

When using libraries like `cva`, some object keys are not used to contain class names.
You can specify which keys will not be parsed by the plugin through this setting.
For example, `cva` has `compoundVariants` and `defaultVariants`.
Note: Since `compoundVariants` can contain class names in its `class` attribute, you can also use callee to ensure that this inner section is parsed while its parent is ignored.

### `config` (default: generated by `tailwindcss/lib/lib/load-config`)

By default, the plugin will try to load the file returned by the official `loadConfig()` tool.

This allows the plugin to use your customized `colors`, `spacing`, `screens`, etc.

You can provide another path or file name for the Tailwind CSS configuration file, such as `"config/tailwind.js"`.

If the external file cannot be loaded (e.g. the path is incorrect or the file has been deleted), an empty object `{}` will be used.

You can also directly inject a pure `object` configuration, such as `{ prefix: "tw-", theme: { ... } }`.

Finally, the plugin will [merge the provided configuration](https://tailwindcss.com/docs/configuration#referencing-in-java-script) with [Tailwind CSS's default configuration](https://github.com/tailwindlabs/tailwindcss/blob/master/stubs/defaultConfig.stub.js).

### `skipClassAttribute` (default: `false`)

Set `skipClassAttribute` to `true` if you only want to check class names inside functions in `callees`.
This will avoid checking `class` and `className` attributes, but will still check `callees` inside these attributes.

### `tags` (default: `[]`)

Optional, if you use tag templates, you should provide them in this array.
